### PR TITLE
RPA level shift and evGW convergence acceleration

### PR DESCRIPTION
**Pull request by baumeier**
Change:
- RPA energy update by gap correction for not explicitly GW corrected levels replaced by update with largest absolute correction in occupied and empty manifolds.

Addition:
- Updating of energies in evGW can now be accelerated using linear mixing or non-linear mixing using the Anderson method. Default is plain update as before.

### CODE REVIEW

File: include/votca/xtp/anderson_mixing.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+#ifndef _VOTCA_XTP_ANDERSON__H
+#define _VOTCA_XTP_ANDERSON__H
+
+#include <iostream>
+#include <memory>
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: not needed

File: include/votca/xtp/anderson_mixing.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+#ifndef _VOTCA_XTP_ANDERSON__H
+#define _VOTCA_XTP_ANDERSON__H
+
+#include <iostream>
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: not needed

File: include/votca/xtp/anderson_mixing.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#include <memory>
+#include <vector>
+#include <votca/xtp/eigen.h>
+namespace votca {
+namespace xtp {
+
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: add a doxygen string to explain what this does

File: include/votca/xtp/anderson_mixing.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+class ANDERSON {
+ public:
+  Eigen::VectorXd NPAndersonMixing(const double alpha);
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: what does NP stand for?

Comment (JensWehner) [reply include/votca/xtp/anderson_mixing.h]: const double alpha is not necessary. double will do. What does it return?

File: include/votca/xtp/anderson_mixing.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#define _VOTCA_XTP_ANDERSON__H
+
+#include <iostream>
+#include <memory>
+#include <vector>
+#include <votca/xtp/eigen.h>
+namespace votca {
+namespace xtp {
+
+class ANDERSON {
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: classes are typically not all capital letters.

File: include/votca/xtp/anderson_mixing.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ public:
+  Eigen::VectorXd NPAndersonMixing(const double alpha);
+
+  void UpdateOutput(const Eigen::VectorXd &newOutput);
+  void UpdateInput(const Eigen::VectorXd &newInput);
+  void SetOrder(const Index max_history) { _max_history = max_history + 1; };
+  bool Info() { return success; }
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: method should be const?

File: include/votca/xtp/anderson_mixing.h (hunk 4 above)
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: shoudl this be const?

File: include/votca/xtp/anderson_mixing.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+namespace votca {
+namespace xtp {
+
+class ANDERSON {
+ public:
+  Eigen::VectorXd NPAndersonMixing(const double alpha);
+
+  void UpdateOutput(const Eigen::VectorXd &newOutput);
+  void UpdateInput(const Eigen::VectorXd &newInput);
+  void SetOrder(const Index max_history) { _max_history = max_history + 1; };
```
Comment (JensWehner) [inline include/votca/xtp/anderson_mixing.h]: why is the method called order, but the input is max_history?

File: include/votca/xtp/gw.h [hunk 8]
```diff
@@ -85,6 +87,7 @@ class GW {
 
  private:
   Index _qptotal;
+  Index _mixing_order;
```
Comment (JensWehner) [inline include/votca/xtp/gw.h]: why add it in the options and as a member variable?

File: src/libxtp/anderson_mixing.cc [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    // Building Linear System for Coefficients
+    const Index used_history = _output.size() - 1;
+    Eigen::MatrixXd A = Eigen::MatrixXd::Zero(used_history, used_history);
+    Eigen::VectorXd c = Eigen::VectorXd::Zero(used_history);
+
+    for (Index m = 1; m <= used_history; m++) {
+
+      c(m - 1) =
+          (DeltaN - _output.at(used_history - m) + _input.at(used_history - m))
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: using the .at operator is normally a bad idea instead use [] operator.

File: src/libxtp/anderson_mixing.cc [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ */
+
+#include <boost/format.hpp>
+#include <votca/xtp/anderson_mixing.h>
+
+namespace votca {
+namespace xtp {
+
+void ANDERSON::UpdateOutput(const Eigen::VectorXd &newOutput) {
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: paper reference?

File: src/libxtp/anderson_mixing.cc [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    const Index used_history = _output.size() - 1;
+    Eigen::MatrixXd A = Eigen::MatrixXd::Zero(used_history, used_history);
+    Eigen::VectorXd c = Eigen::VectorXd::Zero(used_history);
+
+    for (Index m = 1; m <= used_history; m++) {
+
+      c(m - 1) =
+          (DeltaN - _output.at(used_history - m) + _input.at(used_history - m))
+              .cwiseProduct(DeltaN)
+              .sum();
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: this is a .dot product.

File: src/libxtp/anderson_mixing.cc [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+      for (Index j = 1; j <= used_history; j++) {
+        A(m - 1, j - 1) =
+            (DeltaN - _output.at(used_history - m) +
+             _input.at(used_history - m))
+                .cwiseProduct((DeltaN - _output.at(used_history - j) +
+                               _input.at(used_history - j)))
+                .sum();
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: same here

File: src/libxtp/anderson_mixing.cc [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  Index size = _output.size();
+  if (size > _max_history - 1) {
+    _input.erase(_input.begin());
+  }
+  _input.push_back(newInput);
+}
+
+Eigen::VectorXd ANDERSON::NPAndersonMixing(const double alpha) {
+
+  _iteration++;
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: can you not get the iteration cound from the size of input and output? Do you need an iteration counter?

File: src/libxtp/gwbse/gw.cc [hunk 14]
```diff
@@ -48,6 +49,7 @@ void GW::configure(const options& opt) {
   _sigma->configure(sigma_opt);
   _Sigma_x = Eigen::MatrixXd::Zero(_qptotal, _qptotal);
   _Sigma_c = Eigen::MatrixXd::Zero(_qptotal, _qptotal);
+  _mixing_order = _opt.gw_mixing_order;
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: I do not think this is necessary you have value in the _opt struct.

File: src/libxtp/gwbse/gw.cc [hunk 15]
```diff
@@ -168,19 +174,56 @@ void GW::CalculateGWPerturbation() {
...
+        _mixing.UpdateInput(mixed_frequencies);
+        _rpa.UpdateRPAInputEnergies(_dft_energies, mixed_frequencies,
+                                    _opt.qpmin);
+        frequencies = mixed_frequencies;
+
+        for (int i = 0; i < mixed_frequencies.size(); i++) {
+          XTP_LOG(Log::debug, _log)
+              << "... GWSC iter " << i_gw << " state " << i << " "
+              << std::setprecision(9) << mixed_frequencies(i) << std::flush;
+        }
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: can be moved after the if else ?

File: src/libxtp/gwbse/gwbse.cc [hunk 16]
```diff
@@ -319,10 +319,11 @@ void GWBSE::Initialize(tools::Property& options) {
       _gwopt.g_sc_max_iterations);  // convergence criteria for qp iteration
                                     // [Hartree]]
 
-  _gwopt.gw_sc_max_iterations = options.ifExistsReturnElseReturnDefault<Index>(
-      key + ".gw_sc_max_iterations",
-      _gwopt.gw_sc_max_iterations);  // convergence criteria for qp iteration
-                                     // [Hartree]]
+  if (mode == "evGW") {
```
Comment (JensWehner) [inline src/libxtp/gwbse/gwbse.cc]: why this change?

File: src/libxtp/gwbse/gw.cc [hunk 17]
```diff
@@ -168,19 +174,56 @@ void GW::CalculateGWPerturbation() {
         << TimeStamp() << " Calculated screening via RPA" << std::flush;
     XTP_LOG(Log::info, _log)
         << TimeStamp() << " Solving QP equations " << std::flush;
+    if (_opt.gw_mixing_order > 0 && i_gw == 1) {
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: why equals 1?

File: include/votca/xtp/anderson_mixing.h (hunk 4 above)
Comment (JensWehner) [reply include/votca/xtp/anderson_mixing.h]: alpha should be past in the set order/configure method.

Comment (JensWehner) [reply include/votca/xtp/anderson_mixing.h]: and as this is an action, please turn the function name into a verb.

File: src/libxtp/anderson_mixing.cc [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    // Mixing the Potentials
+    for (Index n = 1; n <= used_history; n++) {
+
+      OutMixed += coefficients(n - 1) *
+                  (_output.at(used_history - n) - _output.at(used_history));
+      InMixed += coefficients(n - 1) *
+                 (_input.at(used_history - n) - _input.at(used_history));
+    }
+  }
+
```
Comment (JensWehner) [inline src/libxtp/anderson_mixing.cc]: where is _success used?

File: include/votca/xtp/anderson_mixing.h (hunk 1 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: True.

File: include/votca/xtp/anderson_mixing.h (hunk 2 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: True.

File: include/votca/xtp/anderson_mixing.h (hunk 4 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: Will create a Configure function in place of SetOrder

File: include/votca/xtp/anderson_mixing.h (hunk 5 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: OK.

File: include/votca/xtp/anderson_mixing.h (hunk 6 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: Which one?

File: include/votca/xtp/anderson_mixing.h (hunk 7 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: All changed to order.

File: src/libxtp/anderson_mixing.cc (hunk 10 above)
Comment (baumeier) [reply src/libxtp/anderson_mixing.cc]: Does my diploma thesis suffice? ;)

File: src/libxtp/anderson_mixing.cc (hunk 11 above)
Comment (baumeier) [reply src/libxtp/anderson_mixing.cc]: Jup.

File: src/libxtp/anderson_mixing.cc (hunk 18 above)
Comment (baumeier) [reply src/libxtp/anderson_mixing.cc]: nowhere, going away

File: src/libxtp/anderson_mixing.cc (hunk 13 above)
Comment (baumeier) [reply src/libxtp/anderson_mixing.cc]: That should work as well.

File: include/votca/xtp/gw.h (hunk 8 above)
Comment (baumeier) [reply include/votca/xtp/gw.h]: options alone suffices, indeed

File: src/libxtp/anderson_mixing.cc (hunk 9 above)
Comment (baumeier) [reply src/libxtp/anderson_mixing.cc]: OK

File: src/libxtp/gwbse/gw.cc (hunk 14 above)
Comment (baumeier) [reply src/libxtp/gwbse/gw.cc]: True.

File: include/votca/xtp/anderson_mixing.h (hunk 4 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: Doesn't hurt, I think

File: src/libxtp/gwbse/gw.cc (hunk 15 above)
Comment (baumeier) [reply src/libxtp/gwbse/gw.cc]: Yes.

File: src/libxtp/gwbse/gwbse.cc (hunk 16 above)
Comment (baumeier) [reply src/libxtp/gwbse/gwbse.cc]: Because it seems you could have asked for G0W0, but if you later change gw_sc_max_iterations you override this.

File: src/libxtp/gwbse/gw.cc (hunk 17 above)
Comment (baumeier) [reply src/libxtp/gwbse/gw.cc]: Because I don't think it makes sense to have the dft energies in the history for mixing.

File: src/libxtp/gwbse/gwbse.cc (hunk 16 above)
Comment (JensWehner) [reply src/libxtp/gwbse/gwbse.cc]: okay

File: src/libxtp/gwbse/gw.cc (hunk 17 above)
Comment (JensWehner) [reply src/libxtp/gwbse/gw.cc]: why then >0 instead of ==1?

File: include/votca/xtp/anderson_mixing.h (hunk 3 above)
Comment (baumeier) [reply include/votca/xtp/anderson_mixing.h]: Done

File: src/libxtp/gwbse/gw.cc (hunk 17 above)
Comment (baumeier) [reply src/libxtp/gwbse/gw.cc]: That does work with another change.
